src/igraph_eulerian.c

```diff
@@ -0,0 +1,499 @@
+                igraph_stack_push(&edge_path, curr_e);
+            }
+        }
+    }
+
+    while (!igraph_stack_empty(&edge_path)) {
+        IGRAPH_CHECK(igraph_vector_push_back(res, igraph_stack_pop(&edge_path)));           
+    }
+
+    igraph_stack_destroy(&path);
+    igraph_stack_destroy(&tracker);
+    igraph_stack_destroy(&edge_path);
+    igraph_stack_destroy(&edge_tracker);
+
+    IGRAPH_FINALLY_CLEAN(4);
+
+    return IGRAPH_SUCCESS;
+}
+
+
+int igraph_eulerian_path_directed(igraph_t *graph, igraph_vector_t *res) {
+    igraph_bool_t res_strong, res_weak;
+    igraph_integer_t incoming_excess, outgoing_excess;
+    igraph_integer_t start_node;
+    igraph_t copy;
+    igraph_integer_t i;
+    igraph_vector_t out_degree, in_degree;
+
+    IGRAPH_CHECK(igraph_copy(&copy, graph));
+    IGRAPH_FINALLY(igraph_destroy, &copy);
+
+    incoming_excess = 0;
+    outgoing_excess = 0;
+
+    /* determining the start node */
+    /* also getting the outgoing list for each vector*/
+
+    IGRAPH_CHECK(igraph_vector_init(&out_degree, 0));
+    IGRAPH_FINALLY(igraph_vector_destroy, &out_degree);
+    igraph_degree(graph, &out_degree, igraph_vss_all(), IGRAPH_OUT, IGRAPH_LOOPS);
+
+    IGRAPH_CHECK(igraph_vector_init(&in_degree, 0));
+    IGRAPH_FINALLY(igraph_vector_destroy, &in_degree);
+    igraph_degree(graph, &in_degree, igraph_vss_all(), IGRAPH_IN, IGRAPH_LOOPS);
+
+    /* checking if incoming vertices == outgoing vertices */
+    for (i = 0; i < igraph_vcount(graph); i++) {
+        if (VECTOR(in_degree)[i] != VECTOR(out_degree)[i]) {
+            if ((VECTOR(in_degree)[i] + 1 == VECTOR(out_degree)[i]) && (incoming_excess < 2 && outgoing_excess < 1)) {
+                outgoing_excess++;
+                start_node = i;
+            } else if ((VECTOR(out_degree)[i] + 1 == VECTOR(in_degree)[i]) && (incoming_excess < 1 && outgoing_excess < 2)) {
+                incoming_excess++;
+            } 
+        }
+    }
+
+    IGRAPH_CHECK(igraph_is_connected(graph, &res_strong, IGRAPH_STRONG));
+    IGRAPH_CHECK(igraph_is_connected(graph, &res_weak, IGRAPH_WEAK));
+
+    if ((outgoing_excess == 0 && incoming_excess == 0) && (res_strong)) {
+        start_node = 0;
+        eulerian_path_directed_implementation(&copy, &start_node, &out_degree, res);
+    } else if ((outgoing_excess == 1 && incoming_excess == 1) && 
+        (res_strong || res_weak)) {
+        eulerian_path_directed_implementation(&copy, &start_node, &out_degree, res);
+    }
+
+    igraph_vector_destroy(&in_degree);
+    igraph_vector_destroy(&out_degree);
+    igraph_destroy(&copy);
+    IGRAPH_FINALLY_CLEAN(3);
+
+    return IGRAPH_SUCCESS;
+}
+
+
+int igraph_eulerian_cycle(igraph_t *graph, igraph_vector_t *res) {
+    igraph_bool_t cycle;
+    igraph_bool_t has_path;
+
+    has_path = 0;
+    cycle = 0;
+
+    igraph_is_eulerian(graph, &has_path, &cycle);
+
+    if (!cycle) {
+        IGRAPH_WARNING("Eulerian cycle not possible");
+        return IGRAPH_FAILURE;
```

Comment:
Should be an error, not a warning. Use EINVAL as the error code, as it is more specific than FAILURE. You can check the available error codes in https://github.com/igraph/igraph/blob/master/include/igraph_error.h

Reply:
I did take a look at the potential errors but I was not sure at what the appropriate error code would be in this case. Could you recommend me one?

Reply:
> Could you recommend me one?

I did: EINVAL.

Reply:
I think you misunderstood. Just use `IGRAPH_ERROR` and no `return`. The `IGRAPH_ERROR` macro will do the return (see docs, https://igraph.org/c/doc/igraph-Error.html)

P.S. Originally mentioned here:

https://github.com/igraph/igraph/pull/1346#discussion_r397742443

Reply:
This is fixed now